File: VB Net/RotatingTextButton.vb
' This is Rotating Text Button control.

Imports System.ComponentModel
Imports System.Drawing.Drawing2D

Public Class RotatingTextButton
    Inherits Button

#Region "Properties"

    <Browsable(False)>
    Public Overrides Property Text As String = ""

    <Browsable(False)>
    Public Overrides Property TextAlign As ContentAlignment = ContentAlignment.MiddleCenter

    Private WithEvents Colors_ctc As New CustomTextColors()
    <Browsable(True), DesignerSerializationVisibility(DesignerSerializationVisibility.Content), RefreshProperties(RefreshProperties.Repaint), DescriptionAttribute("Expand to set custom text colors for LinearGradientBrush and PathGradientBrush.")>
    Public Property ButtonTextCustomColors() As CustomTextColors
        Get
            Return Colors_ctc
        End Get
        Set(value As CustomTextColors)
            Colors_ctc = value
            Invalidate()
        End Set
    End Property

    Private m_text As String = "RotatingTextButton"
    <Browsable(True), RefreshProperties(RefreshProperties.All), Description("Text to show on the button."), DefaultValue("RotatingTextButton")>
    Public Property ButtonText As String
        Get
            Return m_text
        End Get
        Set(value As String)
            If m_text <> value Then
                m_text = value
                Invalidate()
            End If
        End Set
    End Property

    Private m_btnBckGradient As Boolean
    <Browsable(True), RefreshProperties(RefreshProperties.All), Description("Enable button back color gradient."), DefaultValue(False)>
    Public Property ButtonBackColorGradient As Boolean
        Get
            Return m_btnBckGradient
        End Get
        Set(value As Boolean)
            If m_btnBckGradient <> value Then
                m_btnBckGradient = value
                Invalidate()
            End If
        End Set
    End Property

    Enum Direction
        Normal
        VerticalLeft
        UpsideDown
        VerticalRight
    End Enum

    Public m_Angle As Integer
    Public m_Dir As Direction = Direction.Normal
    <Browsable(True), RefreshProperties(RefreshProperties.All), Description("The direction of the text."), DefaultValue(Direction.Normal)>
    Public Property ButtonTextDirection() As Direction
        Get
            Return m_Dir
        End Get
        Set(value As Direction)
            If m_Dir <> value Then
                m_Dir = value

                If m_Dir = Direction.Normal OrElse m_Dir = Direction.VerticalRight Then
                    m_Angle = 0
                Else
                    m_Angle = 180
                End If

                Invalidate()
            End If
        End Set
    End Property

    Enum Brush
        SolidBrush = 0
        LinearGradientBrush = 1
        PathGradientBrush = 2
    End Enum

    Private m_brush As Brush = Brush.PathGradientBrush
    <Browsable(True), RefreshProperties(RefreshProperties.All), Description("The color gradient of the text."), DefaultValue(Brush.PathGradientBrush)>
    Public Property ButtonTextBrush As Brush
        Get
            Return m_brush
        End Get
        Set(value As Brush)
            If m_brush <> value Then
                m_brush = value
                Invalidate()
            End If
        End Set
    End Property

    Private m_alpha As Integer = 100
    <Browsable(True), RefreshProperties(RefreshProperties.All), Description("Opacity value for text gradient colors (valid values 0 to 255)."), DefaultValue(100)>
    Public Property ButtonTextAlpha() As Integer
        Get
            Return m_alpha
        End Get
        Set(value As Integer)
            If value < 0 OrElse value > 255 Then value = 100

            If m_alpha <> value Then
                m_alpha = value
                Invalidate()
            End If
        End Set
    End Property

    Private m_lgAngle As Integer = 45
    <Browsable(True), RefreshProperties(RefreshProperties.All), Description("Button text LinearGradient angle (valid values 0 to 360)."), DefaultValue(45)>
    Public Property ButtonTextLGAngle() As Integer
        Get
            Return m_lgAngle
        End Get
        Set(value As Integer)
            If value < 0 OrElse value > 360 Then value = 45

            If m_lgAngle <> value Then
                m_lgAngle = value
                Invalidate()
            End If
        End Set
    End Property

    Private _Highlightcolor As Color = Color.Green
    Public Property HighlightColor() As Color
        Get
            Return _Highlightcolor
        End Get
        Set(value As Color)
            _Highlightcolor = value
        End Set
    End Property

    Private OriginalBackcolor As Color = Nothing
    Private _Highlight As Boolean
    Public Property Highlight() As Boolean
        Get
            Return _Highlight
        End Get
        Set(value As Boolean)
            If OriginalBackcolor = Nothing Then
                OriginalBackcolor = MyBase.BackColor
            End If

            If value Then
                MyBase.BackColor = _Highlightcolor
            Else
                MyBase.BackColor = OriginalBackcolor
            End If

            _Highlight = value
        End Set
    End Property

#End Region

#Region "Constructor"

    Public Sub New()
        MyBase.New()
        SetStyle(ControlStyles.OptimizedDoubleBuffer Or ControlStyles.AllPaintingInWmPaint Or ControlStyles.UserPaint Or ControlStyles.ResizeRedraw Or ControlStyles.ContainerControl Or ControlStyles.SupportsTransparentBackColor, True)
        MyBase.DoubleBuffered = True
        DoubleBuffered = True
        MyBase.BackColor = Color.Black
        MyBase.ForeColor = Color.GreenYellow
        MyBase.Size = New Size(120, 45)
    End Sub

#End Region

#Region "Private Methods"

    Private Sub CustomTextColors_PropertyChanged(propertyName As String) Handles Colors_ctc.PropertyChanged
        Invalidate()
    End Sub

#End Region

#Region "Protected Methods"

    Protected Overrides Sub OnPaint(e As PaintEventArgs)
        MyBase.OnPaint(e)

        Dim gp As New GraphicsPath
        gp.AddRectangle(ClientRectangle)
        Dim pgBrush As New PathGradientBrush(gp)

        If m_btnBckGradient Then
            pgBrush.CenterColor = Color.FromArgb(180, 255, 255, 255)
            pgBrush.SurroundColors = New Color() {Color.FromArgb(0, MyBase.BackColor)}
            e.Graphics.FillRectangle(pgBrush, ClientRectangle)
        End If

        If (m_text IsNot Nothing AndAlso (String.Compare(m_text, "") <> 0)) Then

            Dim format As New StringFormat()

            If m_Dir = Direction.VerticalLeft OrElse m_Dir = Direction.VerticalRight Then
                format.FormatFlags = StringFormatFlags.DirectionVertical
            End If

            format.LineAlignment = StringAlignment.Center
            format.Alignment = StringAlignment.Center

            e.Graphics.TranslateTransform(ClientRectangle.Width / 2, ClientRectangle.Height / 2)
            e.Graphics.RotateTransform(-CSng(m_Angle))
            e.Graphics.TranslateTransform(-ClientRectangle.Width / 2, -ClientRectangle.Height / 2)

            Dim color1, color2 As Color

            If Colors_ctc.CustomTextColorsEnable Then
                color1 = Colors_ctc.CustomTextColor1
                color2 = Colors_ctc.CustomTextColor2
            Else
                color1 = ForeColor
                color2 = ForeColor
            End If

            If m_brush = Brush.SolidBrush Then
                e.Graphics.DrawString(ButtonText, Font, New SolidBrush(color1), ClientRectangle, format)
                Exit Sub
            ElseIf m_brush = Brush.LinearGradientBrush Then
                Dim lgBrush As New LinearGradientBrush(ClientRectangle, color1, Color.FromArgb(m_alpha, ControlPaint.Light(color2)), m_lgAngle)
                e.Graphics.DrawString(ButtonText, Font, lgBrush, ClientRectangle, format)
            Else
                pgBrush.CenterColor = ControlPaint.Light(color1)
                pgBrush.SurroundColors = {Color.FromArgb(m_alpha, ControlPaint.Light(color2))}
                e.Graphics.DrawString(ButtonText, Font, pgBrush, ClientRectangle, format)
            End If
            e.Graphics.ResetTransform()
        End If

    End Sub

#End Region

    <Serializable(), TypeConverter(GetType(ExpandableObjectConverter)), RefreshProperties(RefreshProperties.Repaint), DesignTimeVisible(True)>
    Public Class CustomTextColors

        Event PropertyChanged(propertyName As String)

        Public Sub New()
        End Sub

        Private m_customColorsEnable As Boolean = False
        <Browsable(True), DesignerSerializationVisibility(DesignerSerializationVisibility.Visible), RefreshProperties(RefreshProperties.All), Description("Enable custom text colors for LinearGradientBrush and PathGradientBrush."), DefaultValue(False)>
        Public Property CustomTextColorsEnable As Boolean
            Get
                Return m_customColorsEnable
            End Get
            Set(value As Boolean)
                If m_customColorsEnable <> value Then
                    m_customColorsEnable = value
                    RaiseEvent PropertyChanged("CustomTextColorsEnable")
                End If
            End Set
        End Property

        Private m_color1 As Color = Color.Turquoise
        <Browsable(True), DesignerSerializationVisibility(DesignerSerializationVisibility.Visible), RefreshProperties(RefreshProperties.All), Description("Button text custom color1."), DefaultValue(GetType(Color), "Turquoise")>
        Public Property CustomTextColor1 As Color
            Get
                Return m_color1
            End Get
            Set(value As Color)
                m_color1 = value
                RaiseEvent PropertyChanged("CustomTextColor1")
            End Set
        End Property

        Private m_color2 As Color = Color.SteelBlue
        <Browsable(True), DesignerSerializationVisibility(DesignerSerializationVisibility.Visible), RefreshProperties(RefreshProperties.All), Description("Button text custom color2."), DefaultValue(GetType(Color), "SteelBlue")>
        Public Property CustomTextColor2 As Color
            Get
                Return m_color2
            End Get
            Set(value As Color)
                m_color2 = value
                RaiseEvent PropertyChanged("CustomTextColor2")
            End Set
        End Property

        Public Overrides Function ToString() As String
            If m_customColorsEnable Then
                Return String.Format("Custom Colors Enabled")
            Else
                Return String.Format("Custom Colors Disabled")
            End If
        End Function

    End Class

End Class

File: C#/RotatingTextButton.cs
// This is Rotating Text Button control.

using System;
using System.Drawing;
using System.Windows.Forms;
using System.ComponentModel;
using System.Drawing.Drawing2D;

public class RotatingTextButton : Button
{
	#region "Constructor"

	public RotatingTextButton() : base()
	{
		SetStyle(ControlStyles.OptimizedDoubleBuffer | ControlStyles.AllPaintingInWmPaint | ControlStyles.UserPaint | ControlStyles.ResizeRedraw | ControlStyles.ContainerControl | ControlStyles.SupportsTransparentBackColor, true);
		base.DoubleBuffered = true;
		DoubleBuffered = true;
		base.BackColor = Color.Black;
		base.ForeColor = Color.GreenYellow;
		base.Size = new Size(120, 45);

		Colors_ctc.PropertyChanged += CustomTextColors_PropertyChanged;
	}

	protected override void Dispose(bool disposing)
	{
		try
		{
			if (disposing)
				Colors_ctc.PropertyChanged -= CustomTextColors_PropertyChanged;
		}
		finally
		{
			base.Dispose(disposing);
		}
	}

	#endregion

	#region "Properties"

	[Browsable(false)]
	public override string Text { get; set; }

	[Browsable(false)]
	public override ContentAlignment TextAlign { get; set; }

	private CustomTextColors Colors_ctc = new CustomTextColors();

	[Browsable(true), DesignerSerializationVisibility(DesignerSerializationVisibility.Content), RefreshProperties(RefreshProperties.Repaint), DescriptionAttribute("Expand to set custom text colors for LinearGradientBrush and PathGradientBrush.")]
	public CustomTextColors ButtonTextCustomColors
	{
		get { return Colors_ctc; }
		set
		{
			Colors_ctc = value;
			Invalidate();
		}
	}

	private string m_text = "RotatingTextButton";
	[Browsable(true), RefreshProperties(RefreshProperties.All), Description("Text to show on the button."), DefaultValue("RotatingTextButton")]
	public string ButtonText
	{
		get { return m_text; }
		set
		{
			if (m_text != value)
			{
				m_text = value;
				Invalidate();
			}
		}
	}

	private bool m_btnBckGradient;
	[Browsable(true), RefreshProperties(RefreshProperties.All), Description("Enable button back color gradient."), DefaultValue(false)]
	public bool ButtonBackColorGradient
	{
		get { return m_btnBckGradient; }
		set
		{
			if (m_btnBckGradient != value)
			{
				m_btnBckGradient = value;
				Invalidate();
			}
		}
	}

	public enum Direction
	{
		Normal,
		VerticalLeft,
		UpsideDown,
		VerticalRight
	}

	public int m_Angle;
	public Direction m_Dir = Direction.Normal;

	[Browsable(true), RefreshProperties(RefreshProperties.All), Description("The direction of the text."), DefaultValue(Direction.Normal)]
	public Direction ButtonTextDirection
	{
		get { return m_Dir; }
		set
		{
			if (m_Dir != value)
			{
				m_Dir = value;

				if (m_Dir == Direction.Normal || m_Dir == Direction.VerticalRight)
					m_Angle = 0;
				else
					m_Angle = 180;

				Invalidate();
			}
		}
	}

	public enum Brush
	{
		SolidBrush = 0,
		LinearGradientBrush = 1,
		PathGradientBrush = 2
	}

	private Brush m_brush = Brush.PathGradientBrush;
	[Browsable(true), RefreshProperties(RefreshProperties.All), Description("The color gradient of the text."), DefaultValue(Brush.PathGradientBrush)]
	public Brush ButtonTextBrush
	{
		get { return m_brush; }
		set
		{
			if (m_brush != value)
			{
				m_brush = value;
				Invalidate();
			}
		}
	}

	private int m_alpha = 100;
	[Browsable(true), RefreshProperties(RefreshProperties.All), Description("Opacity value for text gradient colors (valid values 0 to 255)."), DefaultValue(100)]
	public int ButtonTextAlpha
	{
		get { return m_alpha; }
		set
		{
			if (value < 0 || value > 255)
				value = 100;

			if (m_alpha != value)
			{
				m_alpha = value;
				Invalidate();
			}
		}
	}

	private int m_lgAngle = 45;
	[Browsable(true), RefreshProperties(RefreshProperties.All), Description("Button text LinearGradient angle (valid values 0 to 360)."), DefaultValue(45)]
	public int ButtonTextLGAngle
	{
		get { return m_lgAngle; }
		set
		{
			if (value < 0 || value > 360)
				value = 45;

			if (m_lgAngle != value)
			{
				m_lgAngle = value;
				Invalidate();
			}
		}
	}

	private Color _Highlightcolor = Color.Green;
	public Color HighlightColor
	{
		get { return _Highlightcolor; }
		set { _Highlightcolor = value; }
	}

	private Color OriginalBackcolor;
	private bool _Highlight;
	public bool Highlight
	{
		get { return _Highlight; }
		set
		{
			if (OriginalBackcolor == null)
				OriginalBackcolor = base.BackColor;

			if (value)
				base.BackColor = _Highlightcolor;
			else
				base.BackColor = OriginalBackcolor;

			_Highlight = value;
		}
	}

	#endregion

	#region "Private Methods"

	private void CustomTextColors_PropertyChanged(string propertyName)
	{
		Invalidate();
	}

	#endregion

	#region "Protected Methods"

	protected override void OnPaint(System.Windows.Forms.PaintEventArgs e)
	{
		base.OnPaint(e);

		GraphicsPath gp = new GraphicsPath();
		gp.AddRectangle(ClientRectangle);
		PathGradientBrush pgBrush = new PathGradientBrush(gp);

		if (m_btnBckGradient)
		{
			pgBrush.CenterColor = Color.FromArgb(180, 255, 255, 255);
			pgBrush.SurroundColors = new Color[] { Color.FromArgb(0, base.BackColor) };
			e.Graphics.FillRectangle(pgBrush, ClientRectangle);
		}


		if (m_text != null && (string.Compare(m_text, "") != 0))
		{
			StringFormat format = new StringFormat();

			if (m_Dir == Direction.VerticalLeft || m_Dir == Direction.VerticalRight)
				format.FormatFlags = StringFormatFlags.DirectionVertical;

			format.LineAlignment = StringAlignment.Center;
			format.Alignment = StringAlignment.Center;

			e.Graphics.TranslateTransform(ClientRectangle.Width / 2, ClientRectangle.Height / 2);
			e.Graphics.RotateTransform(-Convert.ToSingle(m_Angle));
			e.Graphics.TranslateTransform(-ClientRectangle.Width / 2, -ClientRectangle.Height / 2);

			Color color1, color2;

			if (Colors_ctc.CustomTextColorsEnable)
			{
				color1 = Colors_ctc.CustomTextColor1;
				color2 = Colors_ctc.CustomTextColor2;
			}
			else
			{
				color1 = ForeColor;
				color2 = ForeColor;
			}

			if (m_brush == Brush.SolidBrush)
			{
				e.Graphics.DrawString(ButtonText, Font, new SolidBrush(color1), ClientRectangle, format);
				return;
			}
			else if (m_brush == Brush.LinearGradientBrush)
			{
				LinearGradientBrush lgBrush = new LinearGradientBrush(ClientRectangle, color1, Color.FromArgb(m_alpha, ControlPaint.Light(color2)), m_lgAngle);
				e.Graphics.DrawString(ButtonText, Font, lgBrush, ClientRectangle, format);
			}
			else
			{
				pgBrush.CenterColor = ControlPaint.Light(color1);
				pgBrush.SurroundColors = new Color[] { Color.FromArgb(m_alpha, ControlPaint.Light(color2)) };
				e.Graphics.DrawString(ButtonText, Font, pgBrush, ClientRectangle, format);
			}
			e.Graphics.ResetTransform();
		}

	}

	#endregion

	[Serializable(), TypeConverter(typeof(ExpandableObjectConverter)), RefreshProperties(RefreshProperties.Repaint), DesignTimeVisible(true)]
	public class CustomTextColors
	{

		public event PropertyChangedEventHandler PropertyChanged;
		public delegate void PropertyChangedEventHandler(string propertyName);

		public CustomTextColors()
		{
		}

		private bool m_customColorsEnable = false;
		[Browsable(true), DesignerSerializationVisibility(DesignerSerializationVisibility.Visible), RefreshProperties(RefreshProperties.All), Description("Enable custom text colors for LinearGradientBrush and PathGradientBrush."), DefaultValue(false)]
		public bool CustomTextColorsEnable
		{
			get { return m_customColorsEnable; }
			set
			{
				m_customColorsEnable = value;

				PropertyChanged?.Invoke("CustomTextColorsEnable");
			}
		}

		private Color m_color1 = Color.Turquoise;
		[Browsable(true), DesignerSerializationVisibility(DesignerSerializationVisibility.Visible), RefreshProperties(RefreshProperties.All), Description("Button text custom color1."), DefaultValue(typeof(Color), "Turquoise")]
		public Color CustomTextColor1
		{
			get { return m_color1; }
			set
			{
				m_color1 = value;

                PropertyChanged?.Invoke("CustomTextColor1");
            }
		}

		private Color m_color2 = Color.SteelBlue;
		[Browsable(true), DesignerSerializationVisibility(DesignerSerializationVisibility.Visible), RefreshProperties(RefreshProperties.All), Description("Button text custom color2."), DefaultValue(typeof(Color), "SteelBlue")]
		public Color CustomTextColor2
		{
			get { return m_color2; }
			set
			{
				m_color2 = value;

                PropertyChanged?.Invoke("CustomTextColor2");
            }
		}

		public override string ToString()
		{
			if (m_customColorsEnable)
				return string.Format("Custom Colors Enabled");
			else
				return string.Format("Custom Colors Disabled");
		}

	}

}
